--- htpdate.c	2011-09-04 04:25:46.000000000 +0300
+++ htpdate.c	2011-09-04 18:33:44.182742000 +0300
@@ -30,6 +30,15 @@
 	http://www.gnu.org/copyleft/gpl.html
 */
 
+/*
+	Modifications by mk@dee.su:
+
+	+ Set the time once able to do so, even if not in first poll cycle
+	+ Buffer overflow prevention when copying Date: header contents
+	+ Added -T <tagfile> option to specify output file with initial time step
+	+ Fixed umask setting
+ */
+
 /* Needed to avoid implicit warnings from strptime */
 #define _GNU_SOURCE
 
@@ -240,7 +249,8 @@
 	/* Receive data from the web server
 	   The return code from recv() is the number of bytes received
 	*/
-	if ( recv(server_s, buffer, BUFFERSIZE, 0) != -1 ) {
+	buffer[BUFFERSIZE-1] = '\0';
+	if ( recv(server_s, buffer, BUFFERSIZE-1, 0) != -1 ) {
 
 		/* Assuming that network delay (server->htpdate) is neglectable,
 		   the received web server time "should" match the local time.
@@ -261,8 +271,8 @@
 			timeofday.tv_usec - when;
 
 		/* Look for the line that contains Date: */
-		if ( (pdate = strstr(buffer, "Date: ")) != NULL ) {
-			strncpy(remote_time, pdate + 11, 24);
+		if ( (pdate = strstr(buffer, "\nDate: ")) != NULL  &&  buffer+BUFFERSIZE - pdate > 12 + 24 ) {
+			strncpy(remote_time, pdate + 12, 24);
 
 			if ( strptime( remote_time, "%d %b %Y %T", &tm) != NULL) {
 				/* Web server timestamps are without daylight saving */
@@ -387,7 +397,7 @@
 	puts("htpdate version "VERSION"\n\
 Usage: htpdate [-046abdhlqstxD] [-i pid file] [-m minpoll] [-M maxpoll]\n\
          [-p precision] [-P <proxyserver>[:port]] [-u user[:group]]\n\
-         <host[:port]> ...\n\n\
+         [-T <timeshiftfile>] <host[:port]> ...\n\n\
   -0    HTTP/1.0 request\n\
   -4    Force IPv4 name resolution only\n\
   -6    Force IPv6 name resolution only\n\
@@ -405,6 +415,7 @@
   -q    query only, don't make time changes (default)\n\
   -s    set time\n\
   -t    turn off sanity time check\n\
+  -T    output time delta after initial shift with -s\n\
   -u    run daemon as user\n\
   -x    adjust kernel clock\n\
   host  web server hostname or ip address (maximum of 16)\n\
@@ -448,7 +459,7 @@
 	signal(SIGHUP, SIG_IGN);
 
 	/* Change the file mode mask */
-	umask(0);
+	umask(022);
 
 	/* Change the current working directory */
 	if ( chdir("/") < 0 ) {
@@ -480,8 +491,25 @@
 }
 
 
+static void writetagfile( char *tagfile, double timedelta ) {
+	FILE *file = fopen ( tagfile, "w" );
+
+	if ( !file )
+		printlog (1, "could not create %s", tagfile);
+
+	else {
+		if ( fprintf( file, "%.3f\n", timedelta ) < 0 )
+			printlog (1, "could not write time delta to %s", tagfile);
+
+		if ( fclose( file ) )
+			printlog (1, "could not successfully close %s", tagfile);
+	}
+}
+
+
 int main( int argc, char *argv[] ) {
 	char				*host = NULL, *proxy = NULL, *proxyport = NULL;
+	char				*tagfile = NULL;
 	char				*port;
 	char				*httpversion = DEFAULT_HTTP_VERSION;
 	char				*pidfile = DEFAULT_PID_FILE;
@@ -500,6 +528,7 @@
 	int					maxsleep = DEFAULT_MAX_SLEEP;
 	int					sleeptime = minsleep;
 	int					sw_uid = 0, sw_gid = 0;
+	int					hassettime = 0;
 	time_t				starttime = 0;
 
 	struct passwd		*pw;
@@ -510,7 +539,7 @@
 
 
 	/* Parse the command line switches and arguments */
-	while ( (param = getopt(argc, argv, "046abdhi:lm:p:qstu:xDM:P:") ) != -1)
+	while ( (param = getopt(argc, argv, "046abdhi:lm:p:qstu:xDM:P:T:") ) != -1)
 	switch( param ) {
 
 		case '0':			/* HTTP/1.0 */
@@ -604,6 +633,9 @@
 			proxyport = DEFAULT_PROXY_PORT;
 			splithostport( &proxy, &proxyport );
 			break;
+		case 'T':
+			tagfile = (char *)optarg;
+			break;
 		case '?':
 			return 1;
 		default:
@@ -758,7 +790,7 @@
 		}
 
 		/* Do I really need to change the time?  */
-		if ( sumtimes || !daemonize ) {
+		if ( sumtimes || !daemonize || setmode == 2 ) {
 			/* If a precision was specified and the time offset is small
 			   (< +-1 second), adjust the time with the value of precision
 			*/
@@ -768,10 +800,21 @@
 			/* Correct the clock, if not in "adjtimex" mode */
 			if ( setclock( timeavg, setmode ) < 0 )
 					printlog( 1, "Time change failed" );
+			/* After first correction, do not step through time, only adjust */
+			else if ( setmode == 2 ) {
+				hassettime = 1;
+				setmode = 1;
+			}
 
 			/* Drop root privileges again */
 			if ( sw_uid ) seteuid( sw_uid );
 
+			/* Create a tag file if requested */
+			if ( tagfile && hassettime ) {
+				writetagfile( tagfile, timeavg );
+				tagfile = NULL;
+			}
+
 			if ( daemonize ) {
 				if ( starttime ) {
 					/* Calculate systematic clock drift */
@@ -814,11 +857,6 @@
 			exit(1);
 	}
 
-	/* After first poll cycle do not step through time, only adjust */
-	if ( setmode != 3 ) {
-		setmode = 1;
-	}
-
 	} while ( daemonize );		/* end of infinite while loop */
 
 	exit(0);
