--- /etc/init.d/cabled	2012-01-23 16:09:22.000000000 +0000
+++ /etc/init.d/cabled.new	2012-01-23 16:09:44.000000000 +0000
@@ -26,7 +26,7 @@
 
         # SSD -u sets up USER and HOME, and sanitizes environment
         ebegin "Starting cabled"
-        start-stop-daemon -S -q -c ${cuser} -p ${pidfile} -bm -x ${daemon} --
+        start-stop-daemon -S -q -u ${cuser} -p ${pidfile} -bm -x ${daemon} --
         eend $?
     else
         ewarn "Skipping cabled startup"
