--- /usr/src/linux/scripts/mkcompile_h	2010-12-26 23:28:56.000000000 +0000
+++ /usr/src/linux/scripts/mkcompile_h.new	2010-12-27 00:01:03.000000000 +0000
@@ -65,7 +65,7 @@
 
   echo \#define LINUX_COMPILE_TIME \"`date +%T`\"
   echo \#define LINUX_COMPILE_BY \"`whoami`\"
-  echo \#define LINUX_COMPILE_HOST \"`hostname | $UTS_TRUNCATE`\"
+  echo \#define LINUX_COMPILE_HOST \"liberte\"
 
   domain=`dnsdomainname 2> /dev/null`
   if [ -z "$domain" ]; then
@@ -73,7 +73,7 @@
   fi
 
   if [ -n "$domain" ]; then
-    echo \#define LINUX_COMPILE_DOMAIN \"`echo $domain | $UTS_TRUNCATE`\"
+    echo \#define LINUX_COMPILE_DOMAIN
   else
     echo \#define LINUX_COMPILE_DOMAIN
   fi
