--- /etc/init.d/kexec	2011-01-19 21:48:58.000000000 +0000
+++ /etc/init.d/kexec.new	2011-01-19 21:58:12.000000000 +0000
@@ -74,20 +74,26 @@
 start() {
 	[ "${LOAD_DURING_SHUTDOWN:-yes}" = "yes" ] && return 0
 
+	KPARAM="${KPARAM} halt alert"
+
 	ebegin "Configuring kexec"
 	load_image
 	eend 0
 }
 
 stop() {
-	[ "${LOAD_DURING_SHUTDOWN:-yes}" != "yes" ] && return 0
+	[ "${LOAD_DURING_SHUTDOWN:-yes}" = "no" ] && return 0
 
 	ebegin "Configuring kexec"
+	if [ "`/sbin/runlevel|/bin/cut -c 3`" = "0" ]; then
+		KPARAM="${KPARAM} halt"
+	else
 	if [ "`/sbin/runlevel|/bin/cut -c 3`" != "6" ]; then
 		einfo "Not rebooting, so disabling"
 		kexec -u
 		return 0
 	fi
+	fi
 
 	load_image
 	eend $?
