--- /etc/conf.d/rc	2010-01-12 01:12:41.000000000 +0000
+++ /etc/conf.d/rc.new	2010-01-12 01:12:16.000000000 +0000
@@ -3,7 +3,7 @@
 # This is the number of tty's used in most of the rc-scripts (like
 # consolefont, numlock, etc ...)
 
-RC_TTY_NUMBER=12
+RC_TTY_NUMBER=4
 
 # Set to "yes" if you want the rc system to try and start services
 # in parallel for a slight speed improvement. NOTE: When RC_PARALLEL_STARTUP
@@ -12,13 +12,13 @@
 # You can stop this from happening on the command line by passing --verbose
 # to the init script or by setting RC_VERBOSE="yes" below.
 
-RC_PARALLEL_STARTUP="no"
+RC_PARALLEL_STARTUP="no"
 
 # Set RC_INTERACTIVE to "yes" and you'll be able to press the I key during
 # boot so you can choose to start specific services. Set to "no" to disable
 # this feature.
 
-RC_INTERACTIVE="yes"
+RC_INTERACTIVE="no"
 
 # Do we allow services to be hotplugged? If not, set to RC_HOTPLUG="no"
 # NOTE: This does not affect anything hotplug/udev related, just the
@@ -43,7 +43,7 @@
 # Example - RC_PLUG_SERVICES="net.wlan !net.*"
 # This allows net.wlan and any service not matching net.* to be plugged.
 
-RC_PLUG_SERVICES=""
+RC_PLUG_SERVICES="net.lo !net.*"
 
 # RC_NET_STRICT_CHECKING allows some flexibility with the 'net' service.
 # The following values are allowed:
@@ -78,7 +78,7 @@
 # This is also used to re-enable init script output for init scripts
 # started or stopped from the command line.
 
-RC_VERBOSE="no"
+RC_VERBOSE="no"
 
 # RC_BOOTLOG will generate a log of the boot messages shown on the console.  
 # Useful for headless machines or debugging.  You need to emerge the 
