--- /sbin/splash-functions.sh	2011-05-05 10:42:44.000000000 +0000
+++ /sbin/splash-functions.sh.new	2011-05-05 10:55:54.000000000 +0000
@@ -26,9 +26,9 @@
 export spl_tmpdir="/lib/splash/tmp"
 export spl_fifo="${spl_cachedir}/.splash"
 export spl_pidfile="${spl_cachedir}/daemon.pid"
-export spl_util="//sbin/splash_util.static"
-export spl_daemon="//sbin/fbsplashd.static"
-export spl_decor="//sbin/fbcondecor_ctl.static"
+export spl_util="//usr/sbin/splash_util"
+export spl_daemon="//usr/sbin/fbsplashd"
+export spl_decor="//usr/sbin/fbcondecor_ctl"
 export spl_bindir="/lib/splash/bin"
 
 # This is the main function which handles all events.
@@ -273,7 +273,7 @@
 splash_cache_prep() {
 	# Mount an in-RAM filesystem at spl_cachedir
 	mount -ns -t "${spl_cachetype}" cachedir "${spl_cachedir}" \
-		-o rw,mode=0644,size="${spl_cachesize}"k
+		-o noexec,nosuid,nodev,mode=0755,size="${spl_cachesize}"k
 
 	retval="$?"
 
@@ -341,8 +341,8 @@
 
 # Returns the current splash mode.
 splash_get_mode() {
-	local ctty="$({spl_bindir}/fgconsole)"
-	local mode="$(${spl_util})"
+	local ctty=$(${spl_bindir}/fgconsole)
+	local mode=$(${spl_util} -c getmode)
 
 	if [ "${mode}" = "silent" ]; then
 		echo "silent"
--- /usr/bin/splash_manager	2011-05-04 01:05:28.000000000 +0000
+++ /usr/bin/splash_manager.new	2011-05-05 03:50:12.000000000 +0000
@@ -22,8 +22,8 @@
 #   different from the one used for the silent tty
 
 # System configuration variables
-spl_util=splash_util.static
-spl_daemon=fbsplashd.static
+spl_util=splash_util
+spl_daemon=fbsplashd
 spl_decor=fbcondecor_ctl
 spl_dir="/lib/splash"
 spl_fifo=${spl_dir}/cache/.splash
