--- /usr/portage/sys-auth/polkit/polkit-0.110.ebuild	2013-03-12 11:31:30.000000000 +0000
+++ /usr/portage/sys-auth/polkit/polkit-0.110.ebuild	2013-03-19 15:30:04.961719707 +0000
@@ -61,6 +61,8 @@
 
 	sed -i -e 's|unix-group:wheel|unix-user:0|' src/polkitbackend/*-default.rules || die #401513
 
+	sed -i 's/\$(EXPAT_LIBS).*/&\n$(LIBJS_LIBS)\\/' src/polkitbackend/Makefile.in
+
 	if has_version '>=dev-lang/spidermonkey-1.8.7'; then
 		sed -i \
 			-e '/mozjs/s:185:187:g' \
