diff -udr proxychains-3.1_p20110225.orig/proxychains/core.c proxychains-3.1_p20110225/proxychains/core.c
--- proxychains-3.1_p20110225.orig/proxychains/core.c	2011-02-25 20:48:33.000000000 +0500
+++ proxychains-3.1_p20110225/proxychains/core.c	2012-08-25 19:02:04.610489009 +0600
@@ -178,7 +178,7 @@
   return size;
 }
 
-static int timed_connect(int sock, const struct sockaddr *addr, unsigned int len)
+static int timed_connect(int sock, const struct sockaddr *addr, socklen_t len)
 {
 	int ret,value,value_len;
  	struct pollfd pfd[1];
diff -udr proxychains-3.1_p20110225.orig/proxychains/core.h proxychains-3.1_p20110225/proxychains/core.h
--- proxychains-3.1_p20110225.orig/proxychains/core.h	2011-02-25 20:48:33.000000000 +0500
+++ proxychains-3.1_p20110225/proxychains/core.h	2012-08-25 19:15:37.785322924 +0600
@@ -90,7 +90,7 @@
 typedef int (*getnameinfo_t) (const struct sockaddr *,
 		socklen_t, char *,
 		socklen_t, char *,
-		socklen_t, unsigned int);
+		socklen_t, int);
 getnameinfo_t true_getnameinfo;
 
 typedef struct hostent *(*gethostbyaddr_t) (const void *, socklen_t, int);
diff -udr proxychains-3.1_p20110225.orig/proxychains/libproxychains.c proxychains-3.1_p20110225/proxychains/libproxychains.c
--- proxychains-3.1_p20110225.orig/proxychains/libproxychains.c	2011-02-25 20:48:33.000000000 +0500
+++ proxychains-3.1_p20110225/proxychains/libproxychains.c	2012-08-25 19:15:14.750610897 +0600
@@ -264,7 +264,7 @@
 
 
 
-int connect (int sock, const struct sockaddr *addr, unsigned int len)
+int connect(int sock, const struct sockaddr *addr, socklen_t len)
 {
 	int socktype=0,optlen=0,flags=0,ret=0;
 	char str[256];
@@ -354,7 +354,7 @@
 int getnameinfo (const struct sockaddr * sa,
 			socklen_t salen, char * host,
 			socklen_t hostlen, char * serv,
-			socklen_t servlen, unsigned int flags)
+			socklen_t servlen, int flags)
 {
 	int ret = 0;
 	if(!init_l)
