--- /usr/portage/media-gfx/splashutils/splashutils-1.5.4.4.ebuild	2012-02-07 22:06:35.000000000 +0000
+++ /usr/portage/media-gfx/splashutils/splashutils-1.5.4.4.ebuild.new	2012-02-07 22:05:51.000000000 +0000
@@ -82,6 +82,7 @@
 
 	cd "${S}"
 	ln -sf "${S}/src" "${WORKDIR}/core"
+	sed -i '\:/etc/mtab\>:d' "${S}"/src/libfbsplash.c
 
 	if ! tc-is-cross-compiler && \
 	   has_version "sys-devel/gcc:$(gcc-version)[vanilla]" ; then
