--- /usr/portage/app-emulation/virtualbox-guest-additions/virtualbox-guest-additions-4.2.4.ebuild	2012-11-21 20:45:43.000000000 +0000
+++ /usr/portage/app-emulation/virtualbox-guest-additions/virtualbox-guest-additions-4.2.4.ebuild	2012-11-21 21:02:28.000000000 +0000
@@ -49,7 +49,7 @@
 
 pkg_setup() {
 		linux-mod_pkg_setup
-		BUILD_PARAMS="KERN_DIR=${KV_DIR} KERNOUT=${KV_OUT_DIR}"
+		BUILD_PARAMS="KERN_DIR=${KV_OUT_DIR} KERNOUT=${KV_OUT_DIR}"
 		enewgroup vboxguest
 		enewuser vboxguest -1 /bin/sh /var/run/vboxguest vboxguest
 }
@@ -83,6 +83,9 @@
 				"${WORKDIR}/${mdir}/${vboxheader}"
 		done
 	done
+
+	# Fix split kernels includes
+	sed -i 's@\<kernelpath\>\(.*/generated\>\)@VBOX_LINUX_SRC\1@' "${S}"/Config.kmk
 }
 
 src_configure() {
@@ -92,6 +95,7 @@
 		--disable-sdl-ttf \
 		--disable-pulse \
 		--disable-alsa \
+		--with-linux=${KV_OUT_DIR} \
 		--build-headless || die "configure failed"
 }
 
