--- /usr/portage/x11-drivers/xf86-video-virtualbox/xf86-video-virtualbox-4.2.4.ebuild	2012-11-21 20:45:53.000000000 +0000
+++ /usr/portage/x11-drivers/xf86-video-virtualbox/xf86-video-virtualbox-4.2.4.ebuild	2012-11-21 21:03:01.000000000 +0000
@@ -50,7 +50,7 @@
 
 pkg_setup() {
 	linux-mod_pkg_setup
-	BUILD_PARAMS="KERN_DIR=${KV_DIR} KERNOUT=${KV_OUT_DIR}"
+	BUILD_PARAMS="KERN_DIR=${KV_OUT_DIR} KERNOUT=${KV_OUT_DIR}"
 
 	python_set_active_version 2
 	python_pkg_setup
@@ -88,6 +88,9 @@
 	if gcc-specs-now ; then
 		epatch "${FILESDIR}/${PN}-link-lazy.patch"
 	fi
+
+	# Fix split kernels includes
+	sed -i 's@\<kernelpath\>\(.*/generated\>\)@VBOX_LINUX_SRC\1@' "${S}"/Config.kmk
 }
 
 src_configure() {
@@ -98,6 +101,7 @@
 		--disable-sdl-ttf \
 		--disable-pulse \
 		--disable-alsa \
+		--with-linux=${KV_OUT_DIR} \
 		--build-headless || die "configure failed"
 	source ./env.sh
 }
