--- /usr/portage/x11-drivers/xf86-video-virtualbox/xf86-video-virtualbox-4.1.22.ebuild	2012-07-25 02:24:36.000000000 +0000
+++ /usr/portage/x11-drivers/xf86-video-virtualbox/xf86-video-virtualbox-4.1.22.ebuild	2012-07-25 02:25:33.000000000 +0000
@@ -86,6 +86,9 @@
 	if gcc-specs-now ; then
 		epatch "${FILESDIR}/${PN}-link-lazy.patch"
 	fi
+
+	# Fix split kernels includes
+	sed -i 's@\<kernelpath\>\(.*/generated\>\)@VBOX_LINUX_SRC\1@' "${S}"/Config.kmk
 }
 
 src_configure() {
@@ -96,6 +99,7 @@
 		--disable-sdl-ttf \
 		--disable-pulse \
 		--disable-alsa \
+		--with-linux=${KV_OUT_DIR} \
 		--build-headless || die "configure failed"
 	source ./env.sh
 }
